### ISSUE TITLE
Add more tests to the argparsing library, covering all possible use cases.

### ISSUE THREAD

**Issue by marcospb19** [vrmiguel/ouch#26]
We have a custom argparsing library called `oof`, it's a subcrate in ouch's workspace, so we have to certify ourselves that argparsing is being done correctly.

**Comment by fabricio7p** [comment]
So in my understanding this oof crate act sorta like "clap" right?

Should we test these [util.rs](https://github.com/vrmiguel/ouch/blob/master/oof/src/util.rs) first?
Let me know

**Comment by marcospb19** [comment]
Yes, `oof` act almost like `clap` we used `clap` before, but found some usage limitations for the intended `ouch` usage, so we had to create our own. It does not contain a lot of features, just one function for flag treatment, plus other helper functions, you can it in action at [src/cli.rs](https://github.com/vrmiguel/ouch/blob/master/src/cli.rs).

The function `trim_double_hyphen` is very simple, and the other one is not used, so we can skip util.rs.

We need to test [src/lib.rs](https://github.com/vrmiguel/ouch/blob/master/oof/src/lib.rs) because the implementation is somewhat complex.

There's already a tests module in that file, so probably we'd have to add a new function making sure that whatever crazy input the user gives, the command/error detected is correct.

**Comment by marcospb19** [comment]
I double checked and there is already a test for each function in the `lib.rs` file, but they do not cover a lot of cases, for example, `test_filter_flags` only calls `filter_flags` once, that is not enough, it's necessary to test the failures too to ensure that the program will crash with the correct error message (by checking the returned `OofError`), not only the successful attempts.

You can use `gen_args` and call `filter_flags`, then pattern match to assert the returned error.

Testing `oof` is already a lot, the next step after this would be to add some tests to [src/cli.rs](https://github.com/vrmiguel/ouch/blob/master/src/cli.rs) to ensure now that `parse_args_from` returns the correct `ParsedArgs` (with both correct command and flags), or the correct errors, again, this would require some creativity.

In summary: you will be trying to break our code :laughing:, hopefully you can make a test that exposes some error in my implementation, if you don't find any, that's good news too.

No pressure tho, I know this is not an easy task, I tried to give as much information as possible. For your first PR here, a tiny commit would already be helpful to improve ouch testing.

**Comment by fabricio7p** [comment]
#27 How about this.
I know this is barely useful but this crate is a big oof for my skills haha

**Comment by marcospb19** [comment]
Nice, that's one of the checks we needed.

I will add some comments and keep this issue open because we need more testing before closing it.

**Comment by fabricio7p** [comment]
What else can we make further testings?

**Comment by marcospb19** [comment]
Hey @fabricio7p!

So, we tested the lib itself, now we need to make sure that our [src/cli.rs](https://github.com/vrmiguel/ouch/blob/master/src/cli.rs) `oof` usage is correct, this will be easier to test than #27.

The main function here is `parse_args_from`, that gets called by `parse_args`.

![image](https://user-images.githubusercontent.com/38900226/119688113-1543e700-be1e-11eb-9413-991e9140e628.png)

The first lines of that function are kinda trustable, `oof::matches_any_arg` does not require testing, it's just a for loop that checks for '=='.

`oof::pop_subcommand` is also already trustful, we need to add checks only for `oof::filter_flags`.

**Comment by marcospb19** [comment]
Maybe you can create a closure called `test_argparsing` that returns nothing, it receiver a &str, calls gen_args, pass to parse_args_from and runs the assert inside of it, in the function marked as #[test].

After you do this, we can consider: 
- Using a macro instead
- Making `gen_args` from `oof` a part of the public API, because it can be useful for people who are using it to test.

**Comment by marcospb19** [comment]
One more thing, @fabricio7p, I discovered that you can actually insert match if guards inside of the `matches!` macro.

![image](https://user-images.githubusercontent.com/38900226/119693233-78377d00-be22-11eb-9d56-41988a202ec3.png)

So I edited your code there, this trick might be helpful for this task. Here's [the code](https://github.com/vrmiguel/ouch/blob/5a2028b117e91cfb722ac588b5d818f66d90c0bc/oof/src/lib.rs#L257-L282).
